text/0066-separate-context-propagation.md

```diff
@@ -0,0 +1,557 @@
+# Context Propagation: A Layered Approach 
+
+* [Motivation](#Motivation)
+* [OpenTelemetry layered architecture](#OpenTelemetry-layered-architecture)
+  * [Cross-Cutting Concerns](#Cross-Cutting-Concerns)
+    * [Observability API](#Observability-API)
+    * [Correlations API](#Correlations-API)
+  * [Context Propagation](#Context-Propagation)
+    * [Context API](#Context-API)
+    * [Propagation API](#Propagation-API)
+* [Prototypes](#Prototypes)
+* [Examples](#Examples)
+  * [Global initialization](#Global-initialization)
+  * [Extracting and injecting from HTTP headers](#Extracting-and-injecting-from-HTTP-headers)
+  * [Simplify the API with automated context propagation](#Simplify-the-API-with-automated-context-propagation)
+  * [Implementing a propagator](#Implementing-a-propagator)
+  * [Implementing a concern](#Implementing-a-concern)
+  * [The scope of current context](#The-scope-of-current-context)
+  * [Referencing multiple contexts](#Referencing-multiple-contexts)
+  * [Falling back to explicit contexts](#Falling-back-to-explicit-contexts)
+* [Internal details](#Internal-details)
+* [FAQ](#faq)
+
+Refactor OpenTelemetry into a set of separate cross-cutting concerns which 
+operate on a shared context propagation mechanism.
+
+# Motivation
+
+This RFC addresses the following topics:
+
+**Separation of concerns**  
+* Remove the Tracer dependency from context propagation mechanisms.
+* Handle user data (Correlations) and observability data (SpanContext, etc) 
+  separately.  
+
+**Extensibility**
+* Allow developers to create new applications for context propagation. For 
+  example: A/B testing, encrypted or authenticated data, and new, experimental 
+  forms of observability.
+
+
+# OpenTelemetry layered architecture
+
+The design of OpenTelemetry is based on the principles of [aspect-oriented 
```

Comment:
is it true though? We explicitly ask to modify code to get telemetry

Reply:
I believe it is true. This is based on the same core principles, and addresses the same design problem - how to model concerns which cannot be encapsulated, because they intertwine with all of the other concerns. So I would like to use the same concepts and terminology as AOP, rather than invent new terms. 

It is true that AOP implementations which have runtime support can avoid code modifications. But this is also true with OpenTelemetry. Annotations, auto-installers, etc can eliminate code writing. I would be happy to see runtimes provide even deeper automated support for OpenTelemetry.